### PR TITLE
feat: add FP8 attention and Autoround examples

### PR DESCRIPTION
**Pull request by wimCCC**
SUMMARY:
- Add Qwen3 FP8 Attention + W4A16 AutoRound examples.
- Fix AutoRound handling of activation-only quantization schemes.

Implementation
- Add Qwen3-8B and Qwen3 MoE examples combining FP8 Attention quantization with W4A16 AutoRound for Linear layers.
- Skip activation-only schemes with `weights=None` when building AutoRound configurations.
- Apply AutoRound post-processing only to AutoRound-targeted modules, preserving metadata from other quantization modifiers.

Result
Qwen3-8b-fp8 attention+MXFP8
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8984|±  |0.0083|
|     |       |strict-match    |     5|exact_match|↑  |0.8984|±  |0.0083|

Qwen/Qwen3-30B-A3B-Instruct-2507-fp8 attention+ MXFP8
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9393|±  |0.0066|
|     |       |strict-match    |     5|exact_match|↑  |0.9371|±  |0.0067|
|piqa |      1|none            |     5|acc        |↑  |0.8030|±  |0.0093|
|     |       |none            |     5|acc_norm   |↑  |0.8025|±  |0.0093|

### CODE REVIEW

File: src/llmcompressor/modifiers/autoround/base.py [hunk 1]
```diff
@@ -498,13 +498,17 @@ def _clear_layer_quantization_metadata(module: torch.nn.Module) -> bool:
...
-            # Respect AutoRound's final layer decision: if a layer is set back to
-            # full precision (bits/act_bits > 8), do not restore legacy LLMC
-            # qparams, otherwise the layer can look quantized again.
-            layer_should_be_quantized = check_to_quantized(module)
-            if not layer_should_be_quantized:
-                _clear_layer_quantization_metadata(module)
-                continue
+            # Apply AutoRound's quantization decision only to its target modules.
+            is_autoround_target = module.__class__.__name__ in self.resolved_targets
+            if is_autoround_target:
```
[the remark this answers is not in the document]
Comment (wimCCC) [reply src/llmcompressor/modifiers/autoround/base.py:546]: done

**Comment by yiliu30** [comment]
@codex review

File: examples/autoround/README.md [hunk 2]
```diff
@@ -68,6 +68,9 @@ The accuracy of the quantized model is configured by tuning-related parameters.
 | `wNa16`             | [qwen3_example](./quantization_wNa16/qwen3_example.py)                    | Multiple cards for `Qwen3-235B-A22B`  |
 | `wNa16`             | [qwen3_example_custom_dataset.py](./quantization_wNa16/qwen3_example_custom_dataset.py) | Using custom calibration datasets |
 | `wNa16` + `FP8KV`   | [llama3_example](./quantization_kv_cache/llama3_example.py)               |                                       |
+| `wNa16` + `FP8 Attention` | [llama3_attention_example.py](./quantization_attention/llama3_attention_example.py) | `Linear` weights tuned by AutoRound (`W4A16`), attention modules statically quantized to per-head FP8 (also applies FP8 kv cache) |
```
[the remark this answers is not in the document]
Comment (wimCCC) [reply examples/autoround/README.md]: done

File: src/llmcompressor/modifiers/autoround/base.py [hunk 3]
```diff
@@ -363,11 +363,10 @@ def on_calibration_end(self, state: State, event: Event, **kwargs):
...
         unquantized_layers = []
 
-        for name, module in wrapped_model.named_modules():
-            if (
-                module.__class__.__name__ in self.resolved_targets
-                and getattr(module, "quantization_scheme", None) is None
-            ):
+        for name, module in match_named_modules(
+            wrapped_model, self.resolved_targets, self.ignore
+        ):
```
[the remark this answers is not in the document]
Comment (wimCCC) [reply src/llmcompressor/modifiers/autoround/base.py]: Right — get_unquantized_layer_names collects target modules with no quantization_scheme, to pass to AutoRound's ignore_layers. If we also filter by self.ignore here, explicitly-ignored targets would be excluded from the loop and never reach ignore_layers, so AutoRound would apply its default scheme to them instead of respecting the ignore rule. _postprocess_qparams is different — it loops over all modules unconditionally, so it still needs resolved_targets + ignore to decide whether a module is an AutoRound target at all. That's why self.ignore was removed here but kept there.

File: examples/autoround/quantization_attention/qwen3_quantize_example.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                targets=["Qwen3Attention"],
+                input_activations=QuantizationArgs(
+                    num_bits=8, type="float", strategy="tensor"
+                ),
+            ),
+        },
+    ),
+    AutoRoundModifier(
+        targets="Linear",
+        scheme="W4A16",
```
Comment (yiliu30) [inline examples/autoround/quantization_attention/qwen3_quantize_example.py]: Please use `MXFP8` for this example as well.

File: examples/autoround/quantization_attention/qwen3_moe_quantize_example.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+model_id = "Qwen/Qwen3-30B-A3B-Instruct-2507"
+model = AutoModelForCausalLM.from_pretrained(model_id, dtype="auto")
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+
+DATASET_ID = "HuggingFaceH4/ultrachat_200k"
+DATASET_SPLIT = "train_sft"
+NUM_CALIBRATION_SAMPLES = 128
+MAX_SEQUENCE_LENGTH = 1024
+ITERS = 0
```
Comment (yiliu30) [inline examples/autoround/quantization_attention/qwen3_moe_quantize_example.py]: Please use the default iterations.

File: examples/autoround/quantization_attention/qwen3_moe_quantize_example.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+from compressed_tensors.quantization import QuantizationArgs, QuantizationScheme
+from datasets import load_dataset
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from llmcompressor import oneshot
+from llmcompressor.modifiers.autoround import AutoRoundModifier, fix_batch_if_needed
+from llmcompressor.modifiers.quantization import QuantizationModifier
+
+model_id = "Qwen/Qwen3-30B-A3B-Instruct-2507"
+model = AutoModelForCausalLM.from_pretrained(model_id, dtype="auto")
```
Comment (yiliu30) [inline examples/autoround/quantization_attention/qwen3_moe_quantize_example.py]:
```suggestion
with load_context():
    model = AutoModelForCausalLM.from_pretrained(model_id, dtype="auto")
```

File: examples/autoround/quantization_attention/qwen3_quantize_example.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+from compressed_tensors.offload import dispatch_model
+from compressed_tensors.quantization import QuantizationArgs, QuantizationScheme
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from llmcompressor import oneshot
+from llmcompressor.modifiers.autoround import AutoRoundModifier
+from llmcompressor.modifiers.quantization import QuantizationModifier
+
+model_id = "Qwen/Qwen3-8B"
+model = AutoModelForCausalLM.from_pretrained(model_id, dtype="auto")
```
Comment (yiliu30) [inline examples/autoround/quantization_attention/qwen3_quantize_example.py]:
```suggestion
with load_context():
    model = AutoModelForCausalLM.from_pretrained(model_id, dtype="auto")
```

File: src/llmcompressor/modifiers/autoround/base.py [hunk 8]
```diff
@@ -670,6 +677,9 @@ def _build_layer_config_for_autoround(
                 raise TypeError(
                     f"Expected QuantizationScheme, got {type(quant_scheme)}"
                 )
+            if quant_scheme.weights is None:
+                # Skip activation-only schemes because AutoRound configures weights.
```
Comment (yiliu30) [inline src/llmcompressor/modifiers/autoround/base.py]: It is ok to skip that case, but auto-round could set activation. Please correct the comments.

File: examples/autoround/quantization_attention/qwen3_quantize_example.py (hunk 4 above)
Comment (wimCCC) [reply examples/autoround/quantization_attention/qwen3_quantize_example.py]: done

File: examples/autoround/quantization_attention/qwen3_moe_quantize_example.py (hunk 5 above)
Comment (wimCCC) [reply examples/autoround/quantization_attention/qwen3_moe_quantize_example.py]: done

File: src/llmcompressor/modifiers/autoround/base.py (hunk 8 above)
Comment (wimCCC) [reply src/llmcompressor/modifiers/autoround/base.py]: Right. Skipping the check of other algorithms by weight is not enough. So use _match_autoround_targets instead.

**Comment by dsikka** [comment]
@yiliu30 is this ready to land?

File: src/llmcompressor/modifiers/autoround/base.py [hunk 9]
```diff
@@ -525,14 +539,19 @@ def _clear_layer_quantization_metadata(module: torch.nn.Module) -> bool:
             return cleared_scheme
 
         # Update offload parameters and remove temporary attributes
+        autoround_targets = self._match_autoround_targets(model)
```
Comment (yiliu30) [inline src/llmcompressor/modifiers/autoround/base.py:542]: Could you please double-check whether this still respects AutoRound's final layer decision?

File: tests/llmcompressor/modifiers/autoround/test_base.py [hunk 10]
```diff
@@ -28,6 +28,41 @@ def __init__(self):
...
+        self.up_proj = nn.Linear(128, 128)
+
+
+class _FakeMlp(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.gate_proj = nn.Linear(128, 128)
+
+
+class _RegexDecoderLayer(nn.Module):
```
Comment (yiliu30) [inline tests/llmcompressor/modifiers/autoround/test_base.py]: How about using an tiny model?

**Review by yiliu30 — COMMENTED** [review]
Sorry for the late review. I’ve left a few comments.

**Comment by yiliu30** [comment]
> @yiliu30 is this ready to land?

Sorry @dsikka, I was tied up with some urgent tasks for our upcoming GPU. I’ve reviewed it again and left a few comments.

**Comment by yiliu30** [comment]
Hi @wimCCC please fix the DCO issue as well.

File: src/llmcompressor/modifiers/autoround/base.py (hunk 9 above)
Comment (wimCCC) [reply src/llmcompressor/modifiers/autoround/base.py:542]: Yes. This still preserves AutoRound’s final per-layer quantization decision.  _match_autoround_taget()  only limits the check_to_quantized()  check to modules owned by AutoRoundModifier.  It does not replace or bypass the check. For every AutoRound target, check_to_quantized(module) is still evaluated after AutoRound finishes. If AutoRound sets both bit and act_bits above 8. _clear_layer_quantization_metadata() is called and the continue prevents any stale LLMC qparams from being restored.

File: tests/llmcompressor/modifiers/autoround/test_base.py (hunk 10 above)
Comment (wimCCC) [reply tests/llmcompressor/modifiers/autoround/test_base.py]: done

**Review by yiliu30 — APPROVED** [review]
LGTM, thanks!
